source_code/main_mcu/src/LOGIC/logic_encryption.c

```diff
@@ -396,3 +399,64 @@ void logic_encryption_ecc256_derive_public_key(uint8_t const* priv_key, ecc256_p
 }
 
 
+/*! \fn     logic_encryption_sha1 truncate(uint8_t const *sha1)
+*   \brief  Truncate sha1 hash value to uint32 (from RFC6238)
+*   \param  hash  The hash to truncate (input)
+*   \return uint32_t truncated hash value
+*/
+static uint32_t logic_encryption_sha1_truncate(uint8_t const *sha1)
+{
+    uint8_t offset = sha1[SHA1_OUTPUT_LEN - 1] & 0xF;
+    uint32_t truncated_value = (sha1[offset]  & 0x7F) << 24;
+    truncated_value |= (sha1[offset + 1] & 0xFF) << 16;
+    truncated_value |= (sha1[offset + 2] & 0xFF) <<  8;
+    truncated_value |= (sha1[offset + 3] & 0xFF);
+    return truncated_value;
+}
+
+/*! \fn     logic_encryption_generate_totp(uint8_t *key, uint8_t key_len, uint8_t num_digits, cust_char_t *str, uint8_t str_len)
+*   \brief  RFC6238 TOTP implementation
+*   \param  key      Input secret key
+*   \param  key_len  Input secret key length
+*   \param  str      Output generated TOTP
+*   \param  str_len  Length of "str"
+*   \return uint32_t Number of seconds remaining until next time step
+*/
+uint32_t logic_encryption_generate_totp(uint8_t *key, uint8_t key_len, uint8_t num_digits, cust_char_t *str, uint8_t str_len)
+{
+    uint8_t hmac_sha1_output[SHA1_OUTPUT_LEN] = { 0x0 };
+    memset(str, 0x00, sizeof(cust_char_t)*str_len);
+    uint32_t code;
+    br_hmac_key_context kc;
```

Comment:
can you also memset the context & key context before function exit?

Reply:
Done.